src/server/pfs/server/driver_file.go

```diff
@@ -300,45 +319,62 @@ func (d *driver) listFile(ctx context.Context, file *pfs.File, paginationMarker
 	return errors.EnsureStack(err)
 }
 
-func (d *driver) listFileReverse(ctx context.Context, filename string, s Source, number int64, cb func(*pfs.FileInfo) error) error {
-	// use a fixed size slice since we know the number of files we want
-	fis := make([]*pfs.FileInfo, number)
-	index := 0
-	if err := s.Iterate(ctx, func(fi *pfs.FileInfo, _ fileset.File) error {
-		if pathIsChild(filename, pfsfile.CleanPath(fi.File.Path)) {
-			// wrap around to the start of the slice
-			if index == int(number) {
-				index = 0
-			}
-			fis[index] = fi
-			index++
-		}
-		return nil
-	}); err != nil {
-		return errors.EnsureStack(err)
+type circularList struct {
+	// a circular buffer of file infos
+	buffer []*pfs.FileInfo
+	// next index to be populated
+	index int
+	// size is the number of file infos in the ring
+	size int
+}
+
+func newCircularList(size int64) *circularList {
+	r := &circularList{}
+	r.buffer = make([]*pfs.FileInfo, size)
+	r.index = 0
+	r.size = 0
+	return r
+}
+
+func (r *circularList) add(fi *pfs.FileInfo) {
+	r.buffer[r.index] = fi
+	r.index++
```

Comment:
We could just have a condition for the size increment, rather than increment then set it back to the length of the buffer. Something like this:
```
r.index++
// if we are at the end of the buffer, wrap around
if r.index == len(r.buffer) {
	r.index = 0
}
if r.size < len(r.buffer) {
	r.size++
}
	
```